### PR TITLE
fix: Dragging and dropping files in to OSX dock icon

### PR DESCRIPTION
**Pull request by 9mm**
## What kind of change does this PR introduce?

Currently when a file is dragged over the Neovide dock icon, it doesn't change appearance (indicating it doesn't accept drag and drop). Furthermore, when the file is dragged over (after allowing extensions) it still didn't accept the drag event so is intending to be fixed too.

* fixes https://github.com/neovide/neovide/issues/696
* partially fixes https://github.com/neovide/neovide/issues/330

There are some other issues that could potentially be fixed as well if someone with Windows wants to help. It's possible that drag and dropping on windows doesn't work for a similar reason?

* potentially fixes https://github.com/neovide/neovide/issues/1796

I actually think maybe this is related too, as thats the purpose of the massive CF type bundle is to say every type thats supported as an "Editor" role: https://github.com/neovide/neovide/issues/1682

## Did this PR introduce a breaking change? 

- No

### CODE REVIEW

**Comment by 9mm** [comment]
I made this PR as a draft before getting it working, so I can save all my notes here and get feedback on best approach, ie: how to handle the cargo bundle not supporting it. Can we just fork that or something, and merge one of those PR's? or is there a better approach

**Comment by 9mm** [comment]
Another maybe better option is in build.rs modify the plist file manually with https://crates.io/crates/plist ? The other package is no longer maintained btw it seems

**Comment by fredizzimo** [comment]
I think we are missing some information here. What does cargo bundle not support? And which PRs are you talking about?

**Comment by 9mm** [comment]
Basically in order to have drag and drop support on OSX, it looks like in the `Info.plist` you need to have a key called `CFBundleDocumentTypes`.

https://stackoverflow.com/a/14527463/794481

I checked the bundled `Neovide.app` and it's missing that item from `Info.plist`

Given that this file isn't anywhere in the repo, I am assuming it's coming from here:

https://github.com/neovide/neovide/blob/main/Cargo.toml#L106-L118

I am assuming `cargo-bundle` is using that to generate the `Info.plist.`

https://github.com/burtonageo/cargo-bundle

^ In their docs, they only support a small variety of keys that can be added. There's several complaints in all the Github issues about how he isn't maintaining it with new types, and theres months old open PR's where he's also not adding new types.

Heres an example of a PR adding a key: https://github.com/burtonageo/cargo-bundle/pull/123

That one doesn't help us just showing that there's no actual way to just write your own keys/values which seems crazy to me.

I also could be missing something obvious here... but otherwise I don't know how we can write the proper keys to the Info.plist, as full MacOS support requires a lot of identifiers which are missing, for example all the right click "open with" context menus, setting a program as a default when opening, etc, are all controlled via that.

**Comment by 9mm** [comment]
https://github.com/burtonageo/cargo-bundle/issues/33

**Comment by fredizzimo** [comment]
Ah, I see. I will search around if there are some alternatives to cargo bundle. The status of the projects says "very early alpha", and if the maintainer doesn't have time to work on it, then we have to look for alternatives.

**Comment by 9mm** [comment]
I only did a rudimentary 5 mins of googling so this may not end up being any good, but maybe "tauri-bundler" which looks to be [based on cargo-bundler](https://internals.rust-lang.org/t/cross-platform-bundling/16773)

It looks like tauri is a full build system which wouldn't be desired, but maybe tauri-bundler can just be used separately

Im trying to see though if it supports arbitrary keys... still searching

**Comment by fredizzimo** [comment]
I checked Alacritty  and they seem to be doing it manually with a makefile and `hdiutil` https://github.com/alacritty/alacritty/blob/master/Makefile

And with the icons and plist included in the repository https://github.com/alacritty/alacritty/tree/master/extra/osx/Alacritty.app/Contents

So, I guess that's one way to go. But not sure if it's the best. 

I will do more searching later.

**Comment by 9mm** [comment]
But yes - whatever you think. This bit is outside my wheelhouse haha

**Comment by fredizzimo** [comment]
Maybe we could use the tool from MacVim https://github.com/macvim-dev/macvim/tree/master/src/MacVim/create-dmg

**Comment by 9mm** [comment]
That might be a great idea, I can certainly say MacVim has dialed in the native Mac experience, that's for sure. It's almost too much how they're dialed in 🤣 For example, they're basically in every single right click menu, context menu, top-bar menu. Not that we need all that but good to know it would be supported.

**Comment by shaunsingh** [comment]
> [burtonageo/cargo-bundle#33](https://github.com/burtonageo/cargo-bundle/issues/33)
> 
> ^ Other people wanting keys added, and him saying he doesn't have time to work on it anymore

I think it's in our best interest to move away from cargo-bundle for macOS at the least. Tauri builder (https://github.com/tauri-apps/tauri/tree/dev/tooling/bundler) at the least has not just support for `Info.plist`, but shipping a `dmg` instead of our current `.app` natively.  I can't speak for other platforms, but I imagine there are additional bug fixes.

If everything looks good, I'll whip up a PR to move the build process over to tauri-builder. Hopefully that should make this PR a bit easier to implement.

**Comment by 9mm** [comment]
I'm good wil tauri bundler but I'll defer to @fredizzimo - it looks like there's slight improvements on Linux + Windows as well, maybe some other issues would be fixed by that (ie, drag and drop on other OS's, etc)

**Comment by fredizzimo** [comment]
I would personally go with the MacVim's tool, since it seems more generic, while the tauri bundler seems specialized for their usecase of packaging tauri applications. 

I'm also not sure if the cargo.toml support is there in the tauri bundler anymore, they mentioned that they refactored into a library, and I can only see the library documentation. So if the support is there, it seems to be undocumented. And their CLI tool only seem to be able to build tauri applications.

But if @shaunsingh, can get that to work, and you figure out how to enable the plist properies you need, then it should be fine. 

But for example it looks like you have to pass the file associations in a settings struct to enable what you want https://github.com/tauri-apps/tauri/blob/6892a8cbc1df39c48c3b83f3700b873cbad086ed/tooling/bundler/src/bundle/macos/app.rs#L221

**Comment by 9mm** [comment]
Ah very good points, well I'm good with Macvim's obviously as well too, it does what we need for sure

**Comment by fredizzimo** [comment]
For that one, I think you could try it yourself, you just need to build a release build, and give the right parameters the path to the executable, the path to the icons and so on.

**Comment by 9mm** [comment]
Awesome, I might have some questions but I'll give it a go

**Comment by fredizzimo** [comment]
Yes the macOS CI runs on this https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md

**Comment by fredizzimo** [comment]
I also just noticed that the official repository for create-dmg is https://github.com/create-dmg/create-dmg

**Comment by 9mm** [comment]
Awesome, I'm workin on it already I'll let you know what i come up with.

**Comment by fredizzimo** [comment]
Btw, if I understand it correctly,this is the source folder macvim use, but you only need a few things from there, for example the plist with our own modifications

https://github.com/macvim-dev/macvim/tree/master/src/MacVim

**Comment by fredizzimo** [comment]
Actually here's the command they use https://github.com/macvim-dev/macvim/blob/1a8aef0b01ce606f17115b3e71d4864f0640d1b2/src/Makefile#L3679-L3694

**Comment by fredizzimo** [comment]
Sorry, after studying more, it seems to be more complex than that :cry: 

The plist needs to be compiled into the executable https://github.com/macvim-dev/macvim/blob/1a8aef0b01ce606f17115b3e71d4864f0640d1b2/src/MacVim/MacVim_xcode8.xcodeproj/project.pbxproj#L261 And that's probably partly what those bundle tools does

**Comment by 9mm** [comment]
Yes I was kind of coming to that conclusion at the same time, it also seems like it's just for dmg's after you already have the .app built

**Comment by fredizzimo** [comment]
Hm. based on the Alacritty makefile, I don't see anything special, it copies the executable to the same folder as the plist and renames it to Alacritty.app. So the Alacritty.app goes into a folder with the following structure https://github.com/alacritty/alacritty/tree/3d7d81c8482eb9465763020a397290765b70b541/extra/osx/Alacritty.app/Contents

Then it creates the dmg from that https://github.com/alacritty/alacritty/blob/3d7d81c8482eb9465763020a397290765b70b541/Makefile

**Comment by 9mm** [comment]
Interesting, well I can make a rough version in a single file, and I will try to cobble it together and then someone perhaps could refine it, but ill get the core parts of it working first, if you think thats good approach

**Comment by fredizzimo** [comment]
Yes, I think that's a good approach, if it works, then we at least have full control of the plist file and everything else ourselves.

**Comment by 9mm** [comment]
@fredizzimo im getting close to a good working version, do you have a recommendation for where this file should live?

Id rather its in the proper spot it will eventually be, but if its just going to be moved entirely to build.yml or some github process I suppose it doesn't matter

**Comment by fredizzimo** [comment]
We have the folders `wix` and `snap`. So perhaps `macos-app`? We also have common assets in the `assets' folder, where the Linux deskop file is. So if it's just a single file, everything could also go in asserts. 

The actual commands will go to build.yml, but ideally there would be a shell script that it calls, which can also be manually run. So perhaps putting that and the plist and other macOS specific stuff in macos-app folder is the best. 

At some point we might want to clean this up and move all of these under the same main folder, with subfolders for the different platforms. So I don't think the actual location matters too much at this point.

**Comment by 9mm** [comment]
yep makes sense, thats pretty much how i had it anyway. Thanks! will have something soon

**Comment by 9mm** [comment]
Ok here we go 😎

I made 2 scripts:

1. `macos-builder/make-icns`
  * this generates a `tiff` from 2 input pngs for the purpose of supporting retina in the dmg installer background. If you just use a standard png it will be blurry, so a properly composed tiff must be used
  * this generates an `.icns` file from the input 1024 sized png that I added. This only needs to be run when the icon is changed
3. `macos-builder/run` this generates both an `app` and a `dmg`

Here's a preview of the `dmg` installer I designed:

image

I moved over all the old Info.plist keys. Now im going to work on the actual issue/problem with file types, but wanted to get this up for review first

Current "to solve" list

- this does not auto-clear out bundled items (and it wont overwrite). I didnt want to add `rm` stuff to script as im sure that will change. both the `app` and `dmg` should be cleared out before running
- potentially using the values from `Cargo.toml` instead of redefining them
- interpolating the [CFBundleShortVersionString](https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleshortversionstring) and [CFBundleVersion](https://developer.apple.com/documentation/bundleresources/information_property_list/cfbundleversion) version numbers into `Info.plist` (I'm not sure the difference but `alacritty` seems to just put a `1` for `CFBundleVersion`, potentially hardcoded

**Comment by 9mm** [comment]
Regarding all the copyright stuff, I made everything by hand using existing resources. The font is `JetBrainsMono Nerd Font` Bold Italic so I think that's safe to use

**Comment by 9mm** [comment]
Just as a note, when you run it, and the dmg temporarily opens... don't click it. Its part of the compile process where it mounts the dmg for some reason

```
➜  neovide git:(drag-file-event) macos-builder/run
Created 'Neovide.app' in 'target/release/bundle/osx'
Creating disk image...
...................................................................................................................................
created: /Users/zesty/Code/neovide/target/release/bundle/osx/rw.51273.Neovide.dmg
Mounting disk image...
Device name:     /dev/disk4
Searching for mounted interstitial disk image using /dev/disk4s...
Mount dir:       /Volumes/dmg.NgpJlq
Copying background file 'assets/neovide-dmg-background.tiff'...
Making link to Applications dir...
/Volumes/dmg.NgpJlq
Copying volume icon file 'target/release/bundle/osx/Neovide.app/Contents/Resources/Neovide.icns'...
Running AppleScript to make Finder stuff pretty: /usr/bin/osascript "/var/folders/t8/nccp4pks5xz93n7m_h3mj6mr0000gn/T/createdmg.tmp.XXXXXXXXXX.T45dzPQegQ" "dmg.NgpJlq"
waited 1 seconds for .DS_STORE to be created.
Done running the AppleScript...
Fixing permissions...
Done fixing permissions
Skipping blessing on sandbox
Deleting .fseventsd
Unmounting disk image...
"disk4" ejected.
Compressing disk image...
Preparing imaging engine…
Reading Protective Master Boot Record (MBR : 0)…
   (CRC32 $F76F4199: Protective Master Boot Record (MBR : 0))
Reading GPT Header (Primary GPT Header : 1)…
   (CRC32 $1B77B0BA: GPT Header (Primary GPT Header : 1))
Reading GPT Partition Data (Primary GPT Table : 2)…
   (CRC32 $A0D8ACC9: GPT Partition Data (Primary GPT Table : 2))
Reading  (Apple_Free : 3)…
   (CRC32 $00000000:  (Apple_Free : 3))
Reading disk image (Apple_HFS : 4)…
.............................................................................................................................................................................................
   (CRC32 $58985474: disk image (Apple_HFS : 4))
Reading  (Apple_Free : 5)…
.............................................................................................................................................................................................
   (CRC32 $00000000:  (Apple_Free : 5))
Reading GPT Partition Data (Backup GPT Table : 6)…
.............................................................................................................................................................................................
   (CRC32 $A0D8ACC9: GPT Partition Data (Backup GPT Table : 6))
Reading GPT Header (Backup GPT Header : 7)…
.............................................................................................................................................................................................
   (CRC32 $74C20A2D: GPT Header (Backup GPT Header : 7))
Adding resources…
.............................................................................................................................................................................................
Elapsed Time:  3.943s
File size: 11259856 bytes, Checksum: CRC32 $6DCF9612
Sectors processed: 108621, 60644 compressed
Speed: 7.5MB/s
Savings: 79.8%
created: /Users/zesty/Code/neovide/target/release/bundle/osx/Neovide.dmg
hdiutil does not support internet-enable. Note it was removed in macOS 10.15.
Disk image done
```

**Comment by 9mm** [comment]
File types have been added! So everything now is passing in. I think maybe the problem might be the fact its forking now, perhaps its not properly passing the item to the 2nd process, so im going to find a way to test the fork code with this code, and maybe both should be merged together.

If its not forking, then ill figure out what it is

**Comment by 9mm** [comment]
Ok, that was the problem... when I use the `NEOVIDE_FORK=0` env variable, it now gives an error message instead of just disappearing.

Now the problem seems to be this: https://github.com/neovide/neovide/issues/1682

ill look more into that later but i feel very good with progress so far

image

**Comment by 9mm** [comment]
Hey @fredizzimo I have hit a temporary roadblock. From what I gather it seems I need to implement `NSapplication` using `cocoa` crate.

https://stackoverflow.com/a/2490050/794481

I see some talk about the exact error message here (in first issue comment how he explains 'the application cannot handle this file')

https://github.com/rust-windowing/winit/issues/1751

Currently this looks unresolved... I'm not sure if there's a way to do this manually outside?

If you have any code examples from other open source projects I could potentially use as an example i can just try to get something working

**Comment by 9mm** [comment]
I did see this maybe? https://github.com/rust-windowing/winit/issues/1855#issuecomment-786403931

**Comment by fredizzimo** [comment]
@9mm. This is probably out of scope for winit, since it only deals with windows. But based on some quick searches, it looks like we need to design a document based application. https://developer.apple.com/documentation/appkit/documents_data_and_pasteboard/developing_a_document-based_app/

But since, I just did a very quick search, it's possible that I'm completely wrong.

**Comment by fredizzimo** [comment]
@9mm, the thread you linked mentions https://mrmekon.github.io/fruitbasket/fruitbasket/. Which after a short investigation, I think should allow us to get access to the application and listen to events.

**Comment by 9mm** [comment]
Haha, I just saw that in the comment above yours :)

I did start to attempt to use that but gave up after not having any idea what I'm doing. I'll try again

**Comment by fredizzimo** [comment]
I think you just need to create the instance, and use https://mrmekon.github.io/fruitbasket/fruitbasket/struct.FruitApp.html#method.register_callback and https://mrmekon.github.io/fruitbasket/fruitbasket/struct.FruitApp.html#method.register_apple_event

The loop should already be run by winit

**Comment by 9mm** [comment]
[this](https://stackoverflow.com/questions/50996326/why-isnt-applicationopenfile-called-when-the-application-is-already-running) looks helpful even though I have no idea what it means... saving

**Comment by 9mm** [comment]
So I have tried 4 or 5 times throughout the day w/ different attempts. For starters I don't do any ios/osx dev, or really even any native app work with loops/threads, only web architecture with threads which is very different than dealing with OS level resources, windows, etc

I would say that my impression is that Fruitbasket has to try and boot itself as a separate app or something. I didn't really get their examples working.

I did also just try a regular cocoa based approach based on some example I found, and when I placed it before the loop it said I can't do that before initializing the loop (winit error), and when i placed it after event loop it would just not show the window.

I'll try again but Im' not sure how much further I will get, I have spent a good 7 or 8 hours trying different things or even make some small progress but have not achieved that yet.

**Comment by 9mm** [comment]
Also more info about using fruitbasket as a document type handler

https://github.com/mrmekon/fruitbasket/issues/31

https://github.com/woelper/oculante/issues/33

This better explains what I have experienced despite being out of my league, all my attempts eluded to something like this, and I just don't know enough how to handle multiple loops, etc

> My main problem is that fruitbasket brings its own event loop, and the library I am using (notan) also has its own even loop (receiving input, drawing stuff on screen). I believe that I can't have both. I tried putting fruitbasket's loop in a background thread to listen for file events and then send something to oculante, but I believe fruitbasket only works on the main thread. This is why only on Mac, I start fruitbasket's loop, wait as long as I get a file sent to open, then start the app again with the file as an argument. This works "well" for initial opening, but not for opening files after one was already openend, as fruitbasket is not active any more.

**Comment by fredizzimo** [comment]
@9mm, don't worry, the official Apple documentation is very bad, and you are dealing with layers of abstractions on top of it (winit/fruitbasket), which makes everything very hard.

It might not even be possible to do without modifications to winit. So, I think that we need to open an issue there if we can't find the solution. 

I will also read up a bit about this, but I don't know when, probably not in the next few days at least. But maybe someone else is reading this and can help you.

**Comment by 9mm** [comment]
Thanks, I appreciate that. It definitely is painful, even though a bit fun trying to figure things out from just sheer raw intuition and reading.

No rush or worries! That would be cool, if you feel like it - although Im not trying to just open issues then make people work on them.I I think I will feel inspired again soon and try again.

**Comment by abhillman** [comment]
Just an FYI -- this compiles, but does not deliver said functionality on my box as currently authored.

```
% uname -a
Darwin MBP.local 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:05 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6031 arm64
```

**Comment by abhillman** [comment]
Ideally, a change that would enable document handling for macOS might be upstreamed in cargo-bundle (e.g. https://github.com/burtonageo/cargo-bundle/issues/174).

**Comment by 9mm** [comment]
@abhillman 
it doesnt offer said functionality because it's a draft. namely there is no good way using winit to access `application:openFile:`

I'm not an OSX developer so I dont have the skills to find a way for Neovide to hook into the event loop.

If you read the comments you will see theres multiple event loops that need to be handled, there were some possible ways with fruitbasket, etc, but I couldn't get them to work.

I have already done a lot of the heavy lifting which was a PRECURSOR to this which was adding an installer DMG and supporting the necessary `Info.plist` attributes.

But as far as hooking into that event, I'm at a loss.

Which is a bummer because I'd really love this functionality. I work with dozens of CSV's every day on my desktop and i have no easy way to open them. I have to keep macvim on there to use it for that use case... which sux

**Comment by fredizzimo** [comment]
Also,if you read the first messages, you will find the reason why we decided to move away from `cargo bundle`

The [project status](https://github.com/burtonageo/cargo-bundle#status) is
> Very early alpha. Expect the format of the [package.metadata.bundle] section to change, and there is no guarantee of stabili

And that combined with very sporadic development and, and useful pull requests getting left open without any comments.

**Comment by 9mm** [comment]
^ yes that too. TBH it would be really cool to ship this issue eventually as I did spend a lot of time making a nice installer etc, which separates the good OSX from the chaff

heres also what it is stuck on it seems:

https://github.com/rust-windowing/winit/issues/1751

Would definitely be cool to find another way though, as who knows how far away that is

image

**Comment by abhillman** [comment]
> Also,if you read the first messages, you will find the reason why we decided to move away from `cargo bundle`
> 
> The [project status](https://github.com/burtonageo/cargo-bundle#status) is
> 
> > Very early alpha. Expect the format of the [package.metadata.bundle] section to change, and there is no guarantee of stabili
> 
> And that combined with very sporadic development and useful pull requests getting left open without any comments.

Although `cargo-bundle` is in alpha stage, it has thus-far served the needs of `neovide`. It would be a real gift to the rust community and to rust's cross-platform story to add this functionality to `cargo-bundle`.

**Comment by 9mm** [comment]
@abhillman i doubt cargo-bundle will work any time soon given the sheer amount of new keys required, unless they specify some way to allow arbitrary key entry. You can see there are a huge amount of new keys here that added:

https://github.com/neovide/neovide/pull/2191/files#diff-971b2ac9922c3c9ca5b99bb2125e7bfdd78cd3175e349608c4d723f7aba023d8

**Comment by abhillman** [comment]
> @abhillman i doubt cargo-bundle will work any time soon given the sheer amount of new keys required, unless they specify some way to allow arbitrary key entry. You can see there are a huge amount of new keys here that added:
> 
> https://github.com/neovide/neovide/pull/2191/files#diff-971b2ac9922c3c9ca5b99bb2125e7bfdd78cd3175e349608c4d723f7aba023d8

Let's put aside the `cargo-bundle` thing for now! I think however you and other folks think is best is ideal. That said, I played around with this for a bit. The minimal required addition to a `plist` file to allow for a document to be dropped on top of a app bundle is the following:

```plist
<key>CFBundleDocumentTypes</key>
<array>
  <dict>
    <key>LSItemContentTypes</key>
    <array>
      <string>public.data</string>
    </array>
  </dict>
</array>
```

In addition to this piece and other logistics, what remains is registering a message handler for the OS.

**Comment by abhillman** [comment]
Here's a sketch of how to bring this much of the rest of the way.

**Background**

* [`winit`](https://github.com/rust-windowing/winit), the library responsible for creating a window for `neovide`, produces an [`AppDelegate`](https://github.com/rust-windowing/winit/blob/f8b7c4b78f41beb30ca19fe357abdec0d98a6434/src/platform_impl/macos/app_delegate.rs#L50)

* the way that you would get this to work (i.e. dropping documents on an app bundle's icon and calling a method) using Objective-C is by adding the instance method [`application:openFiles:`](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428742-application?language=objc) to a project's `AppDelegate.m`:

```
- (void)application:(NSApplication *)sender openFiles:(NSArray *) fileNames {
    NSLog(@"Files dragged on: %@", fileNames);
}
```

* Objective-C has a runtime that has advanced reflection and related runtime features (i.e. classes can be fetched by name, methods can be added, etc.)

**Sketch of the solution**

```
use objc::runtime::{
    objc_getClass
};

unsafe fn add_url_handler() {
    let class_name = CString::new("WinitApplicationDelegate").expect("CString::new failed");
    let res = objc_getClass(class_name.as_ptr());

    // TODO: use class_addMethod http://sasheldon.com/rust-objc/objc/runtime/fn.class_addMethod.html to add a `application:openFiles:` method to the application delegate

    // TODO: tell neovim to open the file
}
```

**Comment by abhillman** [comment]
For a simple example demonstrating fetching the `ApplicationDelegate` that can be expanded upon as a kind of harness, if desired, which could then be ported to this repo:

```
use std::ffi::CString;
use winit::{
    event_loop::EventLoop,
    window::WindowBuilder
};

use objc::runtime::{
    objc_getClass
};
use winit::event::Event;

unsafe fn add_url_handler() {
    let class_name = CString::new("WinitApplicationDelegate").expect("CString::new failed");
    let res = objc_getClass(class_name.as_ptr());

    // Prints a memory address like 0x6000019d85a0
    println!("{:?}", res);
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    std::fs::write("/tmp/argv.txt", args.join(" ")).unwrap();

    let event_loop = EventLoop::new().unwrap();
    let _window = WindowBuilder::new().build(&event_loop).unwrap();

    let loop_proxy = event_loop.create_proxy();
    loop_proxy.send_event(()).unwrap();

    event_loop.run(|event, target| {
        match event {
            Event::UserEvent(_) => {
                // As-written, this will only be called once. It is called
                // due to the `loop_proxy.send_event(())` call above.
                unsafe { add_url_handler() }
            }
            _ => {}
        }
    }).expect("TODO: panic message");
}
```

In order to use it, one would create a rust project with the proper dependencies, run `cargo bundle`, and modify the resulting app bundle's `Info.plist` per https://github.com/neovide/neovide/pull/2191#issuecomment-1913460232. From there, dropping a document on the binary will result in calling the application delegate's `application:openFiles:` method, if defined.

**Comment by abhillman** [comment]
I'm a novice at Rust and macOS development, so please forgive me in advance if I am speaking in a way that is it all overly basic.

**Comment by 9mm** [comment]
@abhillman awesome! regarding your Info.plist, the only reason all the other keys are required if I remember right is that OSX requires that you explicitly specify every single filetype. Wildcard `*` does not work anymore (it's been awhile since I checked). I have confirmed in a number of other OSX apps and they all do this (specify every single file type), so I followed their lead.

@fredizzimo what do you think of that above? I don't really know rust well enough. That would be awesome if this works tho

Is there a way 2 people can collaborate on a PR so he can actually commit that to my PR?

**Comment by 9mm** [comment]
I can also give this a try momentarily ,my brain is cobwebs on all this though, I'll have a look again today

**Comment by fredizzimo** [comment]
I don't really know what a good approach is, perhaps it would be worth asking the Winit team what their long-term vision is, and if this could be a good short-term workaround?

@9mm, regarding collaboration, you can either give the other person write permissions to your branch, or they can send a PR to your fork targeting your branch, which you can then merge. But you have to be careful with re-basing if you are co-operating on the same branch, it's the safest to always use merges, they will be cleaned up anyway when we finally squash the PR.

**Comment by polachok** [comment]
I sort of implemented this in https://github.com/polachok/neovide/commit/9802ab31cd91773b07bb0e7b38f113355ac1ef2a with some limitations:
* it doesn't work when neovide isn't already open (not sure how to fix this)
* if multiple files are dragged onto icon, only one file is visible (macvim opens them in tabs)
* uses the `WinitApplicationDelegate` hack above

**Comment by 9mm** [comment]
Awesome! I will check it out soon 💯

**Comment by polachok** [comment]
>it doesn't work when neovide isn't already open (not sure how to fix this)

It actually works if I disable forking. Also, this makes "open with>Neovide" menu in finder work. I guess the "openFiles:" request is sent to the original process.

**Comment by 9mm** [comment]
Interesting, thats def a question for @fredizzimo 

Nice work this is super exciting. This is the main feature left thats painful, so Im really stoked!

**Comment by fredizzimo** [comment]
For the forking, we have this issue:
* https://github.com/neovide/neovide/issues/2147

TLDR; 
I propose that we change the defaults to not fork, neither from the GUI nor from the terminal. That aligns better with how most applications behave when launched from the terminal. But it's not a decision I can take by my own.

**Comment by polachok** [comment]
Just opened a draft PR for this: https://github.com/neovide/neovide/pull/2395

**Comment by fredizzimo** [comment]
I guess this can be closed now?

**Comment by 9mm** [comment]
I believe so!
